webhook/resources/rolloutsave/validator.go

```diff
@@ -0,0 +1,82 @@
+package rolloutsave
+
+import (
+	"github.com/pkg/errors"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/engineapi"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+)
+
+type rolloutSaveValidator struct {
+	admission.DefaultValidator
+	ds *datastore.DataStore
+}
+
+func NewValidator(ds *datastore.DataStore) admission.Validator {
+	return &rolloutSaveValidator{ds: ds}
+}
+
+func (v *rolloutSaveValidator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "rolloutsaves",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.RolloutSave{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+		},
+	}
+}
+
+func (v *rolloutSaveValidator) Create(request *admission.Request, newObj runtime.Object) (err error) {
+	defer func() {
+		if err == nil {
+			return
+		}
+		err = werror.NewInvalidError(err.Error(), "")
+	}()
+
+	rolloutSave := newObj.(*longhorn.RolloutSave)
+	if longhornVersion := rolloutSave.Labels[types.GetVersionLabelKey()]; longhornVersion != "" {
+		// Do not validate RolloutSave created by the backup target controller
+		return nil
+	}
+
+	_, err = v.ds.GetSettingValueExisted(types.SettingNameBackupTarget)
+	if err != nil {
+		return err
+	}
+
+	backupTarget, err := v.ds.GetDefaultBackupTargetRO()
+	if err != nil {
+		return err
+	}
+
+	backupTargetClient, err := engineapi.GetBackupTargetClient(backupTarget, v.ds)
+	if err != nil {
+		return err
+	}
+
+	bsRollouts, err := backupTargetClient.RolloutList(backupTargetClient.URL, backupTargetClient.Credential)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list rollouts from %v", backupTargetClient.URL)
+	}
+
+	for bsRolloutName := range bsRollouts {
+		if bsRolloutName == rolloutSave.Name {
+			return errors.Errorf("%v already exists in the %v: %v", rolloutSave.Name, backupTargetClient.URL, bsRollouts)
+		}
+	}
```

Comment:
IMO, this check looks unnecessary...

Reply:
The intention is to avoid overwriting by another cluster.

Reply:
Then it's better to ask the backupstore to check it as the 1st step of the rollout uploading.

Reply:
The backup target controller will be responsible for creating the missing RolloutSave by periodically checking the remote backup target. So I think it will be confusing if we allow users to create the RolloutSave and then error out.

Reply:
OK. It's redundant but maybe user-friendly. I am OK with this.